src/sdbm.3
.\" $Id$
.TH SDBM 3 "1 March 1990"
.SH NAME
sdbm, dbm_open, dbm_prep, dbm_close, dbm_fetch, dbm_store, dbm_delete, dbm_firstkey, dbm_nextkey, dbm_hash, dbm_rdonly, dbm_error, dbm_clearerr, dbm_dirfno, dbm_pagfno \- data base subroutines
.SH SYNOPSIS
.nf
.ft B
#include <sdbm.h>
.sp
typedef struct {
	char *dptr;
	int dsize;
} datum;
.sp
datum nullitem = { NULL, 0 };
.sp
\s-1DBM\s0 *dbm_open(char *file, int flags, int mode)
.sp
\s-1DBM\s0 *dbm_prep(char *dirname, char *pagname, int flags, int mode)
.sp
void dbm_close(\s-1DBM\s0 *db)
.sp
datum dbm_fetch(\s-1DBM\s0 *db, key)
.sp
int dbm_store(\s-1DBM\s0 *db, datum key, datum val, int flags)
.sp
int dbm_delete(\s-1DBM\s0 *db, datum key)
.sp
datum dbm_firstkey(\s-1DBM\s0 *db)
.sp
datum dbm_nextkey(\s-1DBM\s0 *db)
.sp
long dbm_hash(char *string, int len)
.sp
int dbm_rdonly(\s-1DBM\s0 *db)
int dbm_error(\s-1DBM\s0 *db)
dbm_clearerr(\s-1DBM\s0 *db)
int dbm_dirfno(\s-1DBM\s0 *db)
int dbm_pagfno(\s-1DBM\s0 *db)
.ft R
.fi
.SH DESCRIPTION
.IX "database library" sdbm "" "\fLsdbm\fR"
.IX dbm_open "" "\fLdbm_open\fR \(em open \fLsdbm\fR database"
.IX dbm_prep "" "\fLdbm_prep\fR \(em prepare \fLsdbm\fR database"
.IX dbm_close "" "\fLdbm_close\fR \(em close \fLsdbm\fR routine"
.IX dbm_fetch "" "\fLdbm_fetch\fR \(em fetch \fLsdbm\fR database data"
.IX dbm_store "" "\fLdbm_store\fR \(em add data to \fLsdbm\fR database"
.IX dbm_delete "" "\fLdbm_delete\fR \(em remove data from \fLsdbm\fR database"
.IX dbm_firstkey "" "\fLdbm_firstkey\fR \(em access \fLsdbm\fR database"
.IX dbm_nextkey "" "\fLdbm_nextkey\fR \(em access \fLsdbm\fR database"
.IX dbm_hash "" "\fLdbm_hash\fR \(em string hash for \fLsdbm\fR database"
.IX dbm_rdonly "" "\fLdbm_rdonly\fR \(em return \fLsdbm\fR database read-only mode"
.IX dbm_error "" "\fLdbm_error\fR \(em return \fLsdbm\fR database error condition"
.IX dbm_clearerr "" "\fLdbm_clearerr\fR \(em clear \fLsdbm\fR database error condition"
.IX dbm_dirfno "" "\fLdbm_dirfno\fR \(em return \fLsdbm\fR database bitmap file descriptor"
.IX dbm_pagfno "" "\fLdbm_pagfno\fR \(em return \fLsdbm\fR database data file descriptor"
.IX "database functions \(em \fLsdbm\fR"  dbm_open  ""  \fLdbm_open\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_prep  ""  \fLdbm_prep\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_close  ""  \fLdbm_close\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_fetch  ""  \fLdbm_fetch\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_store  ""  \fLdbm_store\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_delete  ""  \fLdbm_delete\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_firstkey  ""  \fLdbm_firstkey\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_nextkey  ""  \fLdbm_nextkey\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_rdonly  ""  \fLdbm_rdonly\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_error  ""  \fLdbm_error\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_clearerr  ""  \fLdbm_clearerr\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_dirfno  ""  \fLdbm_dirfno\fP
.IX "database functions \(em \fLsdbm\fR"  dbm_pagfno  ""  \fLdbm_pagfno\fP
.LP
This package allows an application to maintain a mapping of <key,value> pairs
in disk files.  This is not to be considered a real database system, but is
still useful in many simple applications built around fast retrieval of a data
value from a key.  This implementation uses an external hashing scheme,
called Dynamic Hashing, as described by Per-Aake Larson in BIT 18 (1978) pp.
184-201.  Retrieval of any item usually requires a single disk access.
The application interface is compatible with the
.IR ndbm (3)
library.
.LP
An
.B sdbm
database is kept in two files usually given the extensions
.B \.dir
and
.BR \.pag .
The
.B \.dir
file contains a bitmap representing a forest of binary hash trees, the leaves
of which indicate data pages in the
.B \.pag
file.
.LP
The application interface uses the
.B datum
structure to describe both
.I keys
and
.IR value s.
A
.B datum
specifies a byte sequence of
.I dsize
size pointed to by
.IR dptr .
If you use
.SM ASCII
strings as
.IR key s
or
.IR value s,
then you must decide whether or not to include the terminating
.SM NUL
byte which sometimes defines strings.  Including it will require larger
database files, but it will be possible to get sensible output from a
.IR strings (1)
command applied to the data file.
.LP
In order to allow a process using this package to manipulate multiple
databases, the applications interface always requires a
.IR handle ,
a
.BR "DBM *" ,
to identify the database to be manipulated.  Such a handle can be obtained
from the only routines that do not require it, namely
.BR dbm_open (\|)
or
.BR dbm_prep (\|).
Either of these will open or create the two necessary files.  The
difference is that the latter allows explicitly naming the bitmap and data
files whereas
.BR dbm_open (\|)
will take a base file name and call
.BR dbm_prep (\|)
with the default extensions.
The
.I flags
and
.I mode
parameters are the same as for
.BR open (2).
.LP
To free the resources occupied while a database handle is active, call
.BR dbm_close (\|).
.LP
Given a handle, one can retrieve data associated with a key by using the
.BR dbm_fetch (\|)
routine, and associate data with a key by using the
.BR dbm_store (\|)
routine.
.LP
The values of the
.I flags
parameter for
.BR dbm_store (\|)
can be either
.BR \s-1DBM_INSERT\s0 ,
which will not change an existing entry with the same key, or
.BR \s-1DBM_REPLACE\s0 ,
which will replace an existing entry with the same key.
Keys are unique within the database.
.LP
To delete a key and its associated value use the
.BR dbm_delete (\|)
routine.
.LP
To retrieve every key in the database, use a loop like:
.sp
.nf
.ft B
for (key = dbm_firstkey(db); key.dptr != NULL; key = dbm_nextkey(db))
        ;
.ft R
.fi
.LP
The order of retrieval is unspecified.
.LP
If you determine that the performance of the database is inadequate or
you notice clustering or other effects that may be due to the hashing
algorithm used by this package, you can override it by supplying your
own
.BR dbm_hash (\|)
routine.  Doing so will make the database unintelligable to any other
applications that do not use your specialized hash function.
.sp
.LP
The following macros are defined in the header file:
.IP
.BR dbm_rdonly (\|)
returns true if the database has been opened read\-only.
.IP
.BR dbm_error (\|)
returns true if an I/O error has occurred.
.IP
.BR dbm_clearerr (\|)
allows you to clear the error flag if you think you know what the error
was and insist on ignoring it.
.IP
.BR dbm_dirfno (\|)
returns the file descriptor associated with the bitmap file.
.IP
.BR dbm_pagfno (\|)
returns the file descriptor associated with the data file.
.SH SEE ALSO
.IR open (2).
.SH DIAGNOSTICS
Functions that return a
.B "DBM *"
handle will use
.SM NULL
to indicate an error.
Functions that return an
.B int
will use \-1 to indicate an error.  The normal return value in that case is 0.
Functions that return a
.B datum
will return
.B nullitem
to indicate an error.
.LP
As a special case of
.BR dbm_store (\|),
if it is called with the
.B \s-1DBM_INSERT\s0
flag and the key already exists in the database, the return value will be 1.
.LP
In general, if a function parameter is invalid,
.B errno
will be set to
.BR \s-1EINVAL\s0 .
If a write operation is requested on a read-only database,
.B errno
will be set to
.BR \s-1ENOPERM\s0 .
If a memory allocation (using
.IR malloc (3))
failed,
.B errno
will be set to
.BR \s-1ENOMEM\s0 .
For I/O operation failures
.B errno
will contain the value set by the relevant failed system call, either
.IR read (2),
.IR write (2),
or
.IR lseek (2).
.SH AUTHOR
.IP "Ozan S. Yigit" (oz@nexus.yorku.ca)
.SH BUGS
The sum of key and value data sizes must not exceed
.B \s-1PAIRMAX\s0
(1008 bytes).
.LP
The sum of the key and value data sizes where several keys hash to the
same value must fit within one bitmap page.
.LP
The
.B \.pag
file will contain holes, so its apparent size is larger than its contents.
When copied through the filesystem the holes will be filled.
.LP
The contents of
.B datum
values returned are in volatile storage.  If you want to retain the values
pointed to, you must copy them immediately before another call to this package.
.LP
The only safe way for multiple processes to (read and) update a database at
the same time, is to implement a private locking scheme outside this package
and open and close the database between lock acquisitions.  It is safe for
multiple processes to concurrently access a database read-only.
.SH APPLICATIONS PORTABILITY
For complete source code compatibility with the Berkeley Unix
.IR ndbm (3)
library, the 
.B sdbm.h
header file should be installed in
.BR /usr/include/ndbm.h .
.LP
The
.B nullitem
data item, and the
.BR dbm_prep (\|),
.BR dbm_hash (\|),
.BR dbm_rdonly (\|),
.BR dbm_dirfno (\|),
and
.BR dbm_pagfno (\|)
functions are unique to this package.


ChangeLog
2000-09-29  John Harper  <john@dcs.warwick.ac.uk>

	* librep.spec.in: override infodir in `make install' command

2000-09-28  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.13.1

2000-09-27  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in, config.h.in: check for strncasecmp ()

2000-09-20  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.13

2000-09-19  John Harper  <john@dcs.warwick.ac.uk>

	* build-info: use `$LOCALDOMAIN', then `dnsdomainname', before
	falling back to `domainname'

2000-09-10  Karl M. Hegbloom  <karlheg@debian.org>

	* man/Makefile.in (librep.pdf): New target

	* Makedefs.in (TEXI2PDF): New variable

2000-09-13  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.13pre2

2000-09-08  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: removed hack to look in etc for install-sh

	* man/repl.texi: new chapter `The REPL'

2000-09-05  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: adapted patch from Jacob for weird debian gmp
	header file location. Misc other changes to gmp checks

2000-09-04  John Harper  <john@dcs.warwick.ac.uk>

	* librep.spec.in: added `--with-extra-cflags=-fstrength-reduce'
	to configure. RH force -fno-strength-reduce on i386, this leads
	to too much register pressure in the vm, spilling one of the
	fixed regs..

	* configure.in: version to 0.13pre1

2000-09-01  John Harper  <john@dcs.warwick.ac.uk>

	* librep.spec.in: added `Requires: gdbm >= 1.8' to avoid gdbm
	brain-damage

	* librep.spec.in: install rep_config.h in -devel pkg

2000-08-31  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: create rep_config.h instead of rep.h

2000-08-29    <john@dcs.warwick.ac.uk>

	* librep.spec.in: librep-devel requires exactly @version@ of
	librep package

2000-08-17  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: add -Wpointer-arith when gcc, only add
	-Wno-implicit-int on solaris

2000-08-11  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: an additional option for configure,
	--with-readline-prefix, to specify an alternate path for the
	readline include/library files (from Albert Chin-A-Young
	<china@thewrittenword.com>)

2000-08-03  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in, config.h.in: check for strtoq () as well as
	strtoll ()

2000-08-01  John Harper  <john@dcs.warwick.ac.uk>

	* intl/intl-compat.c, intl/Makefile.in: removed my edits, intl/
	has no differences to standard gettext intl/ now

	* autogen.sh: new script

	* configure.in: removed ranlib check, remove aux-dir decl.
	Added check for <libintl.h> if gettext is in libc

	* etc/regexp.patch, etc/gc-cons: deleted

	* etc/install-sh: moved to top-level directory

	* etc/ltmain.sh, etc/ltconfig, etc/config.sub,
	etc/config.guess: deleted, use libtoolize to create these

2000-07-29  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: signal an error if no gmp and --without-gmp
	wasn't given (the error message says to use --without-gmp if
	that's want is wanted)

2000-07-28  John Harper  <john@dcs.warwick.ac.uk>

	* Makedefs.in: in COMPILE_ENV changed REPEXECDIR to
	REP_DL_LOAD_PATH

	* configure.in: set GMP_LIBS to -lm if not using gmp

2000-07-24  John Harper  <john@dcs.warwick.ac.uk>

	* install-aliases: don't use -h or -H options to grep. Solaris
	grep chokes on -H and neither are needed anyway

	* librep.spec.in: added install-aliases

2000-07-23  John Harper  <john@dcs.warwick.ac.uk>

	* librep.spec.in: added rep_INSTALL_ALIASES

	* configure.in: version 0.13-pre, bumped libcurrent and libage

	* install-aliases: script to extract and create aliases from
	source files (C or Lisp). Installed into $(repcommonexecdir)

2000-07-20  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.12.4, bumped ${librevision}

	* Makedefs.in: set libtool mode explicitly
	* rules.mk.sh: set libtool mode explicitly; include -rpath
	option
	(from David Kaelbling <drk@sgi.com>)

2000-07-15  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: added `--without-gmp' and
	`--disable-continuations' options

	* config.h.in: added HAVE_STRTOLL, WITH_GMP, WITH_CONTINUATIONS

2000-07-13  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: redirect stderr to /dev/null when trying to run
	`aclocal --print-ac-dir'. Apparently this avoids errors if
	aclocal isn't installed (patch from Albert Chin-A-Young
	<china@thewrittenword.com>)

2000-07-12  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.12.3, bumped ${librevision}

2000-07-10  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: just use ${host_cpu} when checking for known
	stack growth directions

	* librep.spec.in: replaced `/usr/info' by `%{_infodir}'

2000-07-06  John Harper  <john@dcs.warwick.ac.uk>

	* Makefile.in, lisp/Makefile.in, man/Makefile.in,
	src/Makefile.in: updated to partially build in a separate
	directory. This still doesn't work for Lisp sources (adapted
	from Martin Pottendorfer <pottendo@utanet.at>)

2000-06-29  John Harper  <john@dcs.warwick.ac.uk>

	* Makefile.in: added repexecdir to installdirs target (for DOC)

2000-06-22  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.12.2, incremented librevision

2000-06-18  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: version 0.12.1, incremented librevision

	* configure.in, config.h.in: check sizeof long and long long
	(if available), from George Lebl <jirka@5z.com>

2000-06-16  John Harper  <john@dcs.warwick.ac.uk>

	* configure.in: fixed typo when checking for shl_load in libdld

	* configure.in: version 0.12

	* configure.in: check libdld for shl_load

2000-06-09  John Harper  <john@dcs.warwick.ac.uk>

	* Makefile.in: added distclean target as alias for realclean

	* configure.in: changed --with-gmp and --with-gdbm to be more
	conventional --with-gmp-prefix and --with-gdbm-prefix

	* configure.in: when extracting dependences from C files using
	GCC, use `-MM' not `-MM -MG' (from Brad Thompson <yak@mit.edu>)


src/sdbm.c
/*
 * sdbm - ndbm work-alike hashed database library
 * based on Per-Aake Larson's Dynamic Hashing algorithms. BIT 18 (1978).
 * author: oz@nexus.yorku.ca
 * status: public domain.
 *
 * core routines
 */

#include "sdbm.h"
#include "sdbm_tune.h"
#include "sdbm_pair.h"

#include <sys/types.h>
#include <sys/stat.h>
#ifdef BSD42
#include <sys/file.h>
#else
#include <fcntl.h>
#include <memory.h>
#endif
#include <errno.h>
#include <string.h>
#include <unistd.h>

#ifdef __STDC__
#include <stddef.h>
#include <stdlib.h>
#endif

#ifndef NULL
#define NULL	0
#endif

/*
 * forward
 */
static int getdbit (SDBM *, long);
static int setdbit (SDBM *, long);
static int getpage (SDBM *, long);
static datum getnext (SDBM *);
static int makroom (SDBM *, long, int);

/*
 * useful macros
 */
#define bad(x)		((x).dptr == NULL || (x).dsize <= 0)
#define exhash(item)	sdbm_hash((item).dptr, (item).dsize)
#define ioerr(db)	((db)->flags |= SDBM_IOERR)

#define OFF_PAG(off)	(long) (off) * PBLKSIZ
#define OFF_DIR(off)	(long) (off) * DBLKSIZ

static long masks[] = {
	000000000000, 000000000001, 000000000003, 000000000007,
	000000000017, 000000000037, 000000000077, 000000000177,
	000000000377, 000000000777, 000000001777, 000000003777,
	000000007777, 000000017777, 000000037777, 000000077777,
	000000177777, 000000377777, 000000777777, 000001777777,
	000003777777, 000007777777, 000017777777, 000037777777,
	000077777777, 000177777777, 000377777777, 000777777777,
	001777777777, 003777777777, 007777777777, 017777777777
};

datum nullitem = {NULL, 0};

SDBM *
sdbm_open(file, flags, mode)
register char *file;
register int flags;
register int mode;
{
	register SDBM *db;
	register char *dirname;
	register char *pagname;
	register int n;

	if (file == NULL || !*file)
		return errno = EINVAL, (SDBM *) NULL;
/*
 * need space for two separate filenames
 */
	n = strlen(file) * 2 + strlen(DIRFEXT) + strlen(PAGFEXT) + 2;

	if ((dirname = malloc((unsigned) n)) == NULL)
		return errno = ENOMEM, (SDBM *) NULL;
/*
 * build the file names
 */
	dirname = strcat(strcpy(dirname, file), DIRFEXT);
	pagname = strcpy(dirname + strlen(dirname) + 1, file);
	pagname = strcat(pagname, PAGFEXT);

	db = sdbm_prep(dirname, pagname, flags, mode);
	free((char *) dirname);
	return db;
}

SDBM *
sdbm_prep(dirname, pagname, flags, mode)
char *dirname;
char *pagname;
int flags;
int mode;
{
	register SDBM *db;
	struct stat dstat;

	if ((db = (SDBM *) malloc(sizeof(SDBM))) == NULL)
		return errno = ENOMEM, (SDBM *) NULL;

        db->flags = 0;
        db->hmask = 0;
        db->blkptr = 0;
        db->keyptr = 0;
/*
 * adjust user flags so that WRONLY becomes RDWR, 
 * as required by this package. Also set our internal
 * flag for RDONLY if needed.
 */
	if (flags & O_WRONLY)
		flags = (flags & ~O_WRONLY) | O_RDWR;

	else if ((flags & 03) == O_RDONLY)
		db->flags = SDBM_RDONLY;
/*
 * open the files in sequence, and stat the dirfile.
 * If we fail anywhere, undo everything, return NULL.
 */
	if ((db->pagf = open(pagname, flags, mode)) > -1) {
		if ((db->dirf = open(dirname, flags, mode)) > -1) {
/*
 * need the dirfile size to establish max bit number.
 */
			if (fstat(db->dirf, &dstat) == 0) {
/*
 * zero size: either a fresh database, or one with a single,
 * unsplit data page: dirpage is all zeros.
 */
				db->dirbno = (!dstat.st_size) ? 0 : -1;
				db->pagbno = -1;
				db->maxbno = dstat.st_size * BYTESIZ;

				(void) memset(db->pagbuf, 0, PBLKSIZ);
				(void) memset(db->dirbuf, 0, DBLKSIZ);
			/*
			 * success
			 */
				return db;
			}
			(void) close(db->dirf);
		}
		(void) close(db->pagf);
	}
	free((char *) db);
	return (SDBM *) NULL;
}

void
sdbm_close(db)
register SDBM *db;
{
	if (db == NULL)
		errno = EINVAL;
	else {
		(void) close(db->dirf);
		(void) close(db->pagf);
		free((char *) db);
	}
}

datum
sdbm_fetch(db, key)
register SDBM *db;
datum key;
{
	if (db == NULL || bad(key))
		return errno = EINVAL, nullitem;

	if (getpage(db, exhash(key)))
		return sdbm_getpair(db->pagbuf, key);

	return ioerr(db), nullitem;
}

int
sdbm_delete(db, key)
register SDBM *db;
datum key;
{
	if (db == NULL || bad(key))
		return errno = EINVAL, -1;
	if (sdbm_rdonly(db))
		return errno = EPERM, -1;

	if (getpage(db, exhash(key))) {
		if (!sdbm_delpair(db->pagbuf, key))
			return -1;
/*
 * update the page file
 */
		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return ioerr(db), -1;

		return 0;
	}

	return ioerr(db), -1;
}

int
sdbm_store(db, key, val, flags)
register SDBM *db;
datum key;
datum val;
int flags;
{
	int need;
	register long hash;

	if (db == NULL || bad(key))
		return errno = EINVAL, -1;
	if (sdbm_rdonly(db))
		return errno = EPERM, -1;

	need = key.dsize + val.dsize;
/*
 * is the pair too big (or too small) for this database ??
 */
	if (need < 0 || need > PAIRMAX)
		return errno = EINVAL, -1;

	if (getpage(db, (hash = exhash(key)))) {
/*
 * if we need to replace, delete the key/data pair
 * first. If it is not there, ignore.
 */
		if (flags == SDBM_REPLACE)
			(void) sdbm_delpair(db->pagbuf, key);
#ifdef SEEDUPS
		else if (sdbm_duppair(db->pagbuf, key))
			return 1;
#endif
/*
 * if we do not have enough room, we have to split.
 */
		if (!sdbm_fitpair(db->pagbuf, need))
			if (!makroom(db, hash, need))
				return ioerr(db), -1;
/*
 * we have enough room or split is successful. insert the key,
 * and update the page file.
 */
		(void) sdbm_putpair(db->pagbuf, key, val);

		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return ioerr(db), -1;
	/*
	 * success
	 */
		return 0;
	}

	return ioerr(db), -1;
}

/*
 * makroom - make room by splitting the overfull page
 * this routine will attempt to make room for SPLTMAX times before
 * giving up.
 */
static int
makroom(db, hash, need)
register SDBM *db;
long hash;
int need;
{
	long newp;
	char twin[PBLKSIZ];
	char *pag = db->pagbuf;
	char *new = twin;
	register int smax = SPLTMAX;

	do {
/*
 * split the current page
 */
		(void) sdbm_splpage(pag, new, db->hmask + 1);
/*
 * address of the new page
 */
		newp = (hash & db->hmask) | (db->hmask + 1);

/*
 * write delay, read avoidence/cache shuffle:
 * select the page for incoming pair: if key is to go to the new page,
 * write out the previous one, and copy the new one over, thus making
 * it the current page. If not, simply write the new page, and we are
 * still looking at the page of interest. current page is not updated
 * here, as dbm_store will do so, after it inserts the incoming pair.
 */
		if (hash & (db->hmask + 1)) {
			if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
			    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
				return 0;
			db->pagbno = newp;
			(void) memcpy(pag, new, PBLKSIZ);
		}
		else if (lseek(db->pagf, OFF_PAG(newp), SEEK_SET) < 0
			 || write(db->pagf, new, PBLKSIZ) < 0)
			return 0;

		if (!setdbit(db, db->curbit))
			return 0;
/*
 * see if we have enough room now
 */
		if (sdbm_fitpair(pag, need))
			return 1;
/*
 * try again... update curbit and hmask as getpage would have
 * done. because of our update of the current page, we do not
 * need to read in anything. BUT we have to write the current
 * [deferred] page out, as the window of failure is too great.
 */
		db->curbit = 2 * db->curbit +
			((hash & (db->hmask + 1)) ? 2 : 1);
		db->hmask |= db->hmask + 1;

		if (lseek(db->pagf, OFF_PAG(db->pagbno), SEEK_SET) < 0
		    || write(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return 0;

	} while (--smax);
/*
 * if we are here, this is real bad news. After SPLTMAX splits,
 * we still cannot fit the key. say goodnight.
 */
#ifdef BADMESS
	(void) write(2, "sdbm: cannot insert after SPLTMAX attempts.\n", 44);
#endif
	return 0;

}

/*
 * the following two routines will break if
 * deletions aren't taken into account. (ndbm bug)
 */
datum
sdbm_firstkey(db)
register SDBM *db;
{
	if (db == NULL)
		return errno = EINVAL, nullitem;
/*
 * start at page 0
 */
	if (lseek(db->pagf, OFF_PAG(0), SEEK_SET) < 0
	    || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
		return ioerr(db), nullitem;
	db->pagbno = 0;
	db->blkptr = 0;
	db->keyptr = 0;

	return getnext(db);
}

datum
sdbm_nextkey(db)
register SDBM *db;
{
	if (db == NULL)
		return errno = EINVAL, nullitem;
	return getnext(db);
}

/*
 * all important binary trie traversal
 */
static int
getpage(db, hash)
register SDBM *db;
register long hash;
{
	register int hbit;
	register long dbit;
	register long pagb;

	dbit = 0;
	hbit = 0;
	while (dbit < db->maxbno && getdbit(db, dbit))
		dbit = 2 * dbit + ((hash & (1 << hbit++)) ? 2 : 1);

	debug(("dbit: %d...", dbit));

	db->curbit = dbit;
	db->hmask = masks[hbit];

	pagb = hash & db->hmask;
/*
 * see if the block we need is already in memory.
 * note: this lookaside cache has about 10% hit rate.
 */
	if (pagb != db->pagbno) { 
/*
 * note: here, we assume a "hole" is read as 0s.
 * if not, must zero pagbuf first.
 */
		if (lseek(db->pagf, OFF_PAG(pagb), SEEK_SET) < 0
		    || read(db->pagf, db->pagbuf, PBLKSIZ) < 0)
			return 0;
		if (!sdbm_chkpage(db->pagbuf))
			return 0;
		db->pagbno = pagb;

		debug(("pag read: %d\n", pagb));
	}
	return 1;
}

static int
getdbit(db, dbit)
register SDBM *db;
register long dbit;
{
	register long c;
	register long dirb;

	c = dbit / BYTESIZ;
	dirb = c / DBLKSIZ;

	if (dirb != db->dirbno) {
		if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
		    || read(db->dirf, db->dirbuf, DBLKSIZ) < 0)
			return 0;
		db->dirbno = dirb;

		debug(("dir read: %d\n", dirb));
	}

	return db->dirbuf[c % DBLKSIZ] & (1 << dbit % BYTESIZ);
}

static int
setdbit(db, dbit)
register SDBM *db;
register long dbit;
{
	register long c;
	register long dirb;

	c = dbit / BYTESIZ;
	dirb = c / DBLKSIZ;

	if (dirb != db->dirbno) {
		if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
		    || read(db->dirf, db->dirbuf, DBLKSIZ) < 0)
			return 0;
		db->dirbno = dirb;

		debug(("dir read: %d\n", dirb));
	}

	db->dirbuf[c % DBLKSIZ] |= (1 << dbit % BYTESIZ);

	if (dbit >= db->maxbno)
		db->maxbno += DBLKSIZ * BYTESIZ;

	if (lseek(db->dirf, OFF_DIR(dirb), SEEK_SET) < 0
	    || write(db->dirf, db->dirbuf, DBLKSIZ) < 0)
		return 0;

	return 1;
}

/*
 * getnext - get the next key in the page, and if done with
 * the page, try the next page in sequence
 */
static datum
getnext(db)
register SDBM *db;
{
	datum key;

	for (;;) {
		db->keyptr++;
		key = sdbm_getnkey(db->pagbuf, db->keyptr);
		if (key.dptr != NULL)
			return key;
/*
 * we either run out, or there is nothing on this page..
 * try the next one... If we lost our position on the
 * file, we will have to seek.
 */
		db->keyptr = 0;
		if (db->pagbno != db->blkptr++)
			if (lseek(db->pagf, OFF_PAG(db->blkptr), SEEK_SET) < 0)
				break;
		db->pagbno = db->blkptr;
		if (read(db->pagf, db->pagbuf, PBLKSIZ) <= 0)
			break;
		if (!sdbm_chkpage(db->pagbuf))
			break;
	}

	return ioerr(db), nullitem;
}
